### PR TITLE
Use MMTk builder API

### PR DESCRIPTION
**Pull request by qinsoon**

### CODE REVIEW

File: mmtk/src/lib.rs [hunk 1]
```diff
@@ -119,23 +121,18 @@ impl VMBinding for OpenJDK {
...
-        #[cfg(feature = "markcompact")]
-        std::env::set_var("MMTK_PLAN", "MarkCompact");
-        #[cfg(feature = "pageprotect")]
-        std::env::set_var("MMTK_PLAN", "PageProtect");
-        #[cfg(feature = "immix")]
-        std::env::set_var("MMTK_PLAN", "Immix");
-        MMTK::new()
+        assert!(!MMTK_INITIALIZED.load(Ordering::Relaxed));
+        let ret = mmtk::memory_manager::gc_init(&BUILDER);
+        MMTK_INITIALIZED.store(true, std::sync::atomic::Ordering::Relaxed);
```
Comment (wks) [inline mmtk/src/lib.rs:134]: ~~This should be `Release` (or `SeqCst` if we want to be more conservative).~~

`Relaxed` should be is OK here.

**Review by wks — CHANGES_REQUESTED** [review]
If appropriate, we may remove the Cargo features that select plans, because we can now select plan from the command line.

There are some `#include` unused.  They may be added by mistake.

The use of `Relaxed` memory order in this PR is correct, but may need explanation.

File: mmtk/src/api.rs [hunk 2]
```diff
@@ -43,14 +43,39 @@ pub unsafe extern "C" fn release_buffer(ptr: *mut Address, length: usize, capaci
...
+    #[cfg(feature = "gencopy")]
+    memory_manager::process(&BUILDER, "plan", "GenCopy");
+    #[cfg(feature = "marksweep")]
+    memory_manager::process(&BUILDER, "plan", "MarkSweep");
+    #[cfg(feature = "markcompact")]
+    memory_manager::process(&BUILDER, "plan", "MarkCompact");
+    #[cfg(feature = "pageprotect")]
+    memory_manager::process(&BUILDER, "plan", "PageProtect");
+    #[cfg(feature = "immix")]
+    memory_manager::process(&BUILDER, "plan", "Immix");
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply mmtk/src/api.rs:63]: Yeah, we still need them. The current CI scripts use those features. But we should be able to remove them very soon after I rewrite the CI scripts.

File: mmtk/src/api.rs [hunk 3]
```diff
@@ -43,14 +43,39 @@ pub unsafe extern "C" fn release_buffer(ptr: *mut Address, length: usize, capaci
...
+}
+
+#[no_mangle]
+pub extern "C" fn openjdk_is_gc_initialized() -> bool {
+    crate::MMTK_INITIALIZED.load(std::sync::atomic::Ordering::Relaxed)
+}
+
+#[no_mangle]
+pub extern "C" fn mmtk_set_heap_size(size: usize) -> bool {
+    memory_manager::process(&BUILDER, "heap_size", size.to_string().as_str())
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply mmtk/src/api.rs]: I have changed this.

File: openjdk/mmtkHeap.cpp [hunk 4]
```diff
@@ -60,6 +60,9 @@ object iterator??!!
 
 //mmtkGCTaskManager* MMTkHeap::_mmtk_gc_task_manager = NULL;
 
+#include <iostream>
+#include <string>
+using namespace std;
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply openjdk/mmtkHeap.cpp]: Yeah. I was trying to convert the heap size to a string in C++, but the OpenJDK's checker warned me that `new` is not allowed. So I moved the code to Rust, and forgot to remove the imports.

File: mmtk/src/api.rs [hunk 5]
```diff
@@ -43,14 +43,39 @@ pub unsafe extern "C" fn release_buffer(ptr: *mut Address, length: usize, capaci
...
+
+    // Make sure that we haven't initialized MMTk (by accident) yet
+    assert!(!crate::MMTK_INITIALIZED.load(Ordering::Relaxed));
+    // Make sure we initialize MMTk here
+    lazy_static::initialize(&SINGLETON);
+}
+
+#[no_mangle]
+pub extern "C" fn openjdk_is_gc_initialized() -> bool {
+    crate::MMTK_INITIALIZED.load(std::sync::atomic::Ordering::Relaxed)
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply mmtk/src/api.rs]: I changed the memory order to `SeqCst`.

**Review by wks — APPROVED** [review]
LGTM

**Pull request merged**